// @pontem/liquidswap-sdk/dist/tsc/sdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SDK = void 0;
const aptos_1 = require("aptos");
const SwapModule_1 = require("./modules/SwapModule");
const ResourcesModule_1 = require("./modules/ResourcesModule");
const LiquidityModule_1 = require("./modules/LiquidityModule");
const constants_1 = require("./constants");
const initialNetworkOptions = {
    nativeToken: '0x1::aptos_coin::AptosCoin',
    modules: {
        Scripts: constants_1.NETWORKS_MODULES.Scripts,
        CoinInfo: constants_1.NETWORKS_MODULES.CoinInfo,
        CoinStore: constants_1.NETWORKS_MODULES.CoinStore,
    },
    resourceAccount: constants_1.RESOURCES_ACCOUNT,
    moduleAccount: constants_1.MODULES_ACCOUNT,
    moduleAccountV05: constants_1.MODULES_V05_ACCOUNT,
    resourceAccountV05: constants_1.RESOURCES_V05_ACCOUNT,
};
class SDK {
    constructor(options) {
        var _a, _b, _c, _d, _e, _f;
        this._networkOptions = initialNetworkOptions;
        if (options.networkOptions) {
            if ((_a = options.networkOptions) === null || _a === void 0 ? void 0 : _a.nativeToken) {
                this._networkOptions.nativeToken = options.networkOptions.nativeToken;
            }
            if ((_b = options.networkOptions) === null || _b === void 0 ? void 0 : _b.modules) {
                this._networkOptions.modules = options.networkOptions.modules;
            }
            if ((_c = options.networkOptions) === null || _c === void 0 ? void 0 : _c.moduleAccount) {
                this._networkOptions.moduleAccount =
                    options.networkOptions.moduleAccount;
            }
            if ((_d = options.networkOptions) === null || _d === void 0 ? void 0 : _d.resourceAccount) {
                this._networkOptions.resourceAccount =
                    options.networkOptions.resourceAccount;
            }
            if ((_e = options.networkOptions) === null || _e === void 0 ? void 0 : _e.moduleAccountV05) {
                this._networkOptions.moduleAccountV05 =
                    options.networkOptions.moduleAccountV05;
            }
            if ((_f = options.networkOptions) === null || _f === void 0 ? void 0 : _f.resourceAccountV05) {
                this._networkOptions.resourceAccountV05 =
                    options.networkOptions.resourceAccountV05;
            }
        }
        this._client = new aptos_1.AptosClient(options.nodeUrl, options.nodeOptions);
        this._swap = new SwapModule_1.SwapModule(this);
        this._resources = new ResourcesModule_1.ResourcesModule(this);
        this._liquidity = new LiquidityModule_1.LiquidityModule(this);
        this._curves = {
            uncorrelated: `${this._networkOptions.moduleAccount}::curves::Uncorrelated`,
            stable: `${this._networkOptions.moduleAccount}::curves::Stable`,
            uncorrelatedV05: `${this._networkOptions.moduleAccountV05}::curves::Uncorrelated`,
            stableV05: `${this._networkOptions.moduleAccountV05}::curves::Stable`,
        };
    }
    get Swap() {
        return this._swap;
    }
    get Resources() {
        return this._resources;
    }
    get Liquidity() {
        return this._liquidity;
    }
    get client() {
        return this._client;
    }
    get networkOptions() {
        return this._networkOptions;
    }
    get curves() {
        return this._curves;
    }
}
exports.SDK = SDK;
