// @thalalabs/surf/build/cjs/hooks/useSubmitTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSubmitTransaction = void 0;
const react_1 = require("react");
const wallet_adapter_react_1 = require("@aptos-labs/wallet-adapter-react");
const useSubmitTransaction = () => {
    const { connected, signAndSubmitTransaction } = (0, wallet_adapter_react_1.useWallet)();
    const [isIdle, setIsIdle] = (0, react_1.useState)(true);
    const [isLoading, setIsLoading] = (0, react_1.useState)(false);
    const [data, setResult] = (0, react_1.useState)();
    const [error, setError] = (0, react_1.useState)();
    const idRef = (0, react_1.useRef)(0);
    function reset() {
        setIsIdle(true);
        setIsLoading(false);
        setResult(undefined);
        setError(undefined);
        idRef.current = idRef.current + 1;
    }
    async function submitTransaction(payload, options) {
        if (connected === false) {
            throw new Error('Wallet is not connected');
        }
        if (signAndSubmitTransaction === undefined) {
            return;
        }
        const id = idRef.current;
        if (!isIdle) {
            throw new Error('Transaction is already in progress');
        }
        setIsLoading(true);
        setIsIdle(false);
        let result;
        try {
            result = await signAndSubmitTransaction({
                data: {
                    ...payload,
                    functionArguments: payload.functionArguments.map((arg) => {
                        if (Array.isArray(arg)) {
                            return arg.map((item) => item);
                        }
                        else if (typeof arg === 'object') {
                            throw new Error(`a value of struct type: ${arg} is not supported`);
                        }
                        else {
                            return arg;
                        }
                    }),
                },
                options: options,
            });
            if (id === idRef.current) {
                setResult(result);
                setIsLoading(false);
            }
        }
        catch (e) {
            setError(error);
            setIsLoading(false);
            throw e;
        }
        return result;
    }
    return { submitTransaction, reset, isLoading, isIdle, data, error };
};
exports.useSubmitTransaction = useSubmitTransaction;
